File: JavaPractice/src/ReadLogFilePassFailurePercentage.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/*
 * This is a snippet of a static log file:

YYYYMMDD_hh:mm:ss::jobname::status::<message>

20170304_16:24:32::foo::PASS::success
20170304_16:48:32::bar::PASS::Yay!
20170304_16:50:42::baz::FAIL::OOM
20170304_16:51:06::job::FAIL::syntax error
20170304_17:03:21::job1::PASS::
20170304_17:27:19::job2::FAIL::halt
20170304_17:30:05::job3::FAIL::pki cert error
20170304_18:01:55::job2::FAIL::
20170304_18:45:22::job3::FAIL::pki cert error

write a function that takes the log file name as an argument, extract all the PASSes and print them out to screen in the following format, make sure it's sorted by date/time:

PASS 95%
    date_time, jobname, message (if any)
    date_time, jobname
    date_time, jobname, message
FAIL 5%
    date_time, jobname, message (if any)
    date_time, jobname
    date_time, jobname, message
 * 
 */

public class ReadLogFilePassFailurePercentage {

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		printPassLoggingInformation("//Users/divyesh.shah/Documents/workspace/JavaPractice/inputFileName.txt");
	}

	static void printPassLoggingInformation(String inputFileNameWithPath) {
        
        if(inputFileNameWithPath == null || inputFileNameWithPath.length() == 0)
            throw new IllegalArgumentException("Pass valid input");
            
        BufferedReader reader;
        
        try {
            
            reader = new BufferedReader(new FileReader(inputFileNameWithPath));
            String currentline = reader.readLine();
            float totalcounter = 0;
            float totalpass = 0;
            List<String> passResult = new ArrayList<>();
            List<String> failResult = new ArrayList<>();
            while(currentline != null) {
                
                String[] wordsplit = currentline.split("::");
                if(wordsplit[2].equals("PASS")) {
                   if(wordsplit.length > 3) {
                        passResult.add(String.format("%s,%s,%s", wordsplit[0],wordsplit[1],wordsplit[3]));
                   }
                    else{
                        passResult.add(String.format("%s,%s", wordsplit[0],wordsplit[1]));
                    }
                    totalpass++;
                }
                else {
                    if(wordsplit.length > 3) {
                        failResult.add(String.format("%s,%s,%s", wordsplit[0],wordsplit[1],wordsplit[3]));
                    }
                    else {
                        failResult.add(String.format("%s,%s", wordsplit[0],wordsplit[1]));
                    }
                }
                totalcounter++;
                /*
                if(wordsplit[3] != " ")
                    System.out.println(String.format("%s,%s,%s", wordsplit[0],wordsplit[1],wordsplit[3]));
                else
                    System.out.println(String.format("%s,%s", wordsplit[0],wordsplit[1]));      
                */
                currentline = reader.readLine();
            }
            
            reader.close();
            
            System.out.println(String.format("%.2f",(totalpass/totalcounter)*100));
            for(String s : passResult)
                System.out.println(s);
            
            System.out.println(String.format("%.2f",(totalcounter - totalpass)/totalcounter * 100));
            for(String s : failResult)
                System.out.println(s);
                
        }
        catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        
    }

}

File: PythonPractice/src/ReadLogFilePassFailurePercentage.py
import os

def print_pass_logging_information(input_file_name_with_path):
    if not input_file_name_with_path or not os.path.isfile(input_file_name_with_path):
        raise ValueError("Pass a valid file path")

    total_counter = 0
    total_pass = 0
    pass_result = []
    fail_result = []

    try:
        with open(input_file_name_with_path, 'r') as file:
            for line in file:
                wordsplit = line.strip().split("::")
                if len(wordsplit) < 4:
                    continue

                date_time, jobname, status, *message = wordsplit
                message = message[0] if message else ""

                if status == "PASS":
                    if message:
                        pass_result.append(f"{date_time},{jobname},{message}")
                    else:
                        pass_result.append(f"{date_time},{jobname}")
                    total_pass += 1
                else:
                    if message:
                        fail_result.append(f"{date_time},{jobname},{message}")
                    else:
                        fail_result.append(f"{date_time},{jobname}")
                total_counter += 1

        pass_result.sort(key=lambda x: x.split(',')[0])
        fail_result.sort(key=lambda x: x.split(',')[0])

        if total_counter > 0:
            print(f"PASS {total_pass / total_counter * 100:.2f}%")
            for result in pass_result:
                print(result)

            print(f"FAIL {(total_counter - total_pass) / total_counter * 100:.2f}%")
            for result in fail_result:
                print(result)
        else:
            print("No entries to process")

    except FileNotFoundError:
        print("File not found")
    except IOError:
        print("Error reading file")
    except Exception as e:
        print(f"An error occurred: {e}")

def main():
    # Example usage
    input_file_name_with_path = "//Users/divyesh.shah/Documents/workspace/JavaPractice/inputFileName.txt"
    print_pass_logging_information(input_file_name_with_path)

if __name__ == "__main__":
    main()
